Scoped nowarn

Martin521 opened the pull request:
## Description

Implements [Scoped Nowarn](https://github.com/fsharp/fslang-suggestions/issues/278) according to draft [RFC FS-1146](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1146-scoped-nowarn.md).

This PR has taken a while. I had to deal with much more complexity than I imagined when I naively volunteered to tackle the feature request. Anyway, here we are.

I have split the PR into 7 commits that can be reviewed in sequence.
All of them compile, 1 and 4 - 7 also pass all tests locally.


1. Add the feature flag, baseline tests, and the core `WarnScopes` module. See `src/Compiler/SyntaxTree/WarnScopes.fsi` and the RFC for the functionality of the module.

2. Add the necessary changes to lexing and parsing. Note that the warn directives can no longer be collected during parsing (since they can now appear not only in top-level modules, but anywhere). So we collect them during lexing, similar to the processing of #if/#else/#endif directives.

3. Remove legacy #nowarn processing (but hold off AST changes)

4. Integrate the WarnScopes functionality and test it

5. Add warn directive trivia (but hold off AST changes)

6. Enable warn directive trivia (which means AST changes)

7. Remove defunct types and parameters related to former #nowarn processing (more AST changes)

There is also a separate commit for the IlVerify baseline updates (change in line numbers only)


## Checklist

- [X] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [X] Release notes entry updated
- [ ] Create documentation update PRs (see RFC)

psfinaki commented:
Hi @Martin521 - thanks for the contribution. It's a substantial effort and we appreciate it. The PR is on our radar - just keep in mind that it's big and specific, and it will take time to find capacity for it.

If anyone from the community gets to thoroughly review it, that would be valuable as well. 

Thanks for your diligence and patience :)

KevinRansom commented on src/Compiler/Driver/CompilerConfig.fs:
```diff
@@ -877,6 +878,14 @@ type TcConfigBuilder =
             compilationMode = TcGlobals.CompilationMode.Unset
         }
 
+    member tcConfigB.SetLangVersion v =
```
Adding a mutator for langVersion is fine but mostly unnecessary, but please don't let the langversion mutator set other values that is not the paradigm we use.
TcConfigBuilder is a mutable record and so modifying values is kind of what it's about.

Martin521 replied:
When `tcConfigB.diagnosticsOptions` is set to `FSharpDiagnosticOptions.Default`, the langversion is not yet known. The langversion comes only much later with `ParseCompilerOpions`.
There are two alternatives that I considered but didn't like.
The first was to add to `tcConfigB` an updated `FSharpDiagnosticOptions` (updated with the langversion) in all the places where `ParseCompilerOptions` is called (in fsc.fs, fsi.fs, BackgroundCompiler.fs, TransparentCompiler.fs and FSharpCheckerResults.fs).
The second was to thread the whole `tcConfig` through all the function calls to the places where I need only the diagnosticsOptions.
But let me check if I find still another way.

Martin521 replied:
I didn't find any new alternative, but the first alternative mentioned above is of course doable. It just means duplication of that logic in six places. Let me know if this is preferred and I will implement it.

KevinRansom commented on src/Compiler/Utilities/range.fs:
```diff
@@ -193,28 +193,29 @@ type FileIndexTable() =
-                    FileSystem.NormalizePathShim filePath
-                else
-                    filePath
+        lock fileToIndexTable
```
It looks like you submitted a significantly better version of this fix that locks only when the filepath is not found on first lookup.  I think there may be a still better solution, because it seems that the indexToFileTable is where the underlying race occurred, and that would grab a lock only when a completely new normalized path is found I think.

KevinRansom reviewed (approved):
I still have a long way to go, this is an interesting feature.

Martin521 commented on src/Compiler/SyntaxTree/WarnScopes.fs:
```diff
@@ -0,0 +1,390 @@
+        ns |> removeQuotes |> Option.bind removePrefix |> Option.bind parseInt
+
+    let private regex =
+        Regex(""" *#(nowarn|warnon|\S+)(?: +([^ \r\n/;]+))*(?:;;)? *(\/\/.*)?$""", RegexOptions.CultureInvariant)
```
This code is never called for `#   nowarn` because of the lexer rule that brings us here.
But a description of the regex makes sense.

Martin521 replied:
I would consider the fact that the compiler accepts spaces after the hash sign a bug.
But it can certainly be replicated in the new version (it has to happen here and in the lex grammar then).
I guess this should become another RFC discussion.

Martin521 replied:
I have added the regex description.

Regarding the changed behavior, personally, I wouldn't allow a space after the hash. But if the team decides we should, I can implement it as mentioned above. Note that this syntax exists in only 5 unique pieces of code in the whole of github.

Martin521 commented on tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.release.bsl:
```diff
@@ -2236,9 +2236,6 @@ FSharp.Compiler.CodeAnalysis.FSharpParsingOptions: FSharp.Compiler.CodeAnalysis.
 FSharp.Compiler.CodeAnalysis.FSharpParsingOptions: FSharp.Compiler.CodeAnalysis.FSharpParsingOptions get_Default()
 FSharp.Compiler.CodeAnalysis.FSharpParsingOptions: FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions DiagnosticOptions
 FSharp.Compiler.CodeAnalysis.FSharpParsingOptions: FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions get_DiagnosticOptions()
-FSharp.Compiler.CodeAnalysis.FSharpParsingOptions: Int32 CompareTo(FSharp.Compiler.CodeAnalysis.FSharpParsingOptions)
```
These methods are no longer auto-generated because the record has now a `obj` member and thus is no longer comparable.

Martin521 commented on tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.release.bsl:
```diff
@@ -2897,8 +2893,12 @@ FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions: Microsoft.FSharp.Collection
+FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions: Microsoft.FSharp.Core.FSharpOption`1[System.Object] WarnScopeData@
+FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions: Microsoft.FSharp.Core.FSharpOption`1[System.Object] get_WarnScopeData()
 FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions: System.String ToString()
-FSharp.Compiler.Diagnostics.FSharpDiagnosticOptions: Void .ctor(Int32, Boolean, Microsoft.FSharp.Collections.FSharpList`1[System.Int32], Microsoft.FSharp.Collections.FSharpList`1[System.Int32], Microsoft.FSharp.Collections.FSharpList`1[System.Int32], Microsoft.FSharp.Collections.FSharpList`1[System.Int32])
```
These methods are no longer auto-generated because FSharpDiagnosticsOptions has now a obj member and thus is no longer comparable.

Martin521 commented on tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.release.bsl:
```diff
@@ -6417,20 +6407,6 @@ FSharp.Compiler.Syntax.QualifiedNameOfFile: Int32 get_Tag()
 FSharp.Compiler.Syntax.QualifiedNameOfFile: System.String Text
 FSharp.Compiler.Syntax.QualifiedNameOfFile: System.String ToString()
 FSharp.Compiler.Syntax.QualifiedNameOfFile: System.String get_Text()
-FSharp.Compiler.Syntax.ScopedPragma: Boolean Equals(FSharp.Compiler.Syntax.ScopedPragma)
```
This one is no longer needed

Martin521 commented on tests/fsharp/core/load-script/ProjectDriver.fsx:
```diff
@@ -1,9 +1,10 @@
 // #Conformance #FSI 
 #load "ThisProject.fsx"
+#nowarn "44"
```
This is implementing the RFC, so it should be discussed over there.
See also [this comment](https://github.com/fsharp/fslang-design/discussions/786#discussioncomment-10716719).

Martin521 replied:
I do think that this new behavior for scripts should be under the feature flag. This still needs to be fixed.

Martin521 replied:
Done

Martin521 replied:
So now, as specified in the RFC, for preview we have the new behavior (#nowarn is effective only locally), while for langVersion <= 9.0 we have the old behavior (#nowarn in scripts is effective for the whole project). I don't know if / how we can make this test langVersion specific, that's why the additional #nowarn is still in.

Martin521 commented on tests/service/data/SyntaxTree/Attribute/RangeOfAttribute.fs.bsl:
```diff
@@ -30,4 +30,5 @@ ImplFile
            Expr (Do (Const (Unit, (3,3--3,5)), (3,0--3,5)), (3,0--3,5))],
           PreXmlDocEmpty, [], None, (2,0--4,0), { LeadingKeyword = None })],
       (true, true), { ConditionalDirectives = []
+                      WarnDirectives = []
```
That's why I spent effort to create the successive commits. Almost all of the baseline changes are introduced only in commit 6. (See the PR description above)

Martin521 commented on tests/FSharp.Compiler.ComponentTests/CompilerDirectives/NonStringArgs.fs:
```diff
@@ -43,42 +43,7 @@ module NonStringArgs =
-        |> shouldFail
-        |> withDiagnostics [
-            if languageVersion = "8.0" then
-                (Error 3350, Line 4, Col 5, Line 4, Col 7, "Feature '# directives with non-quoted string arguments' is not available in F# 8.0. Please use language version 9.0 or greater.")
```
This is implementing the RFC, which assumes that compiler directives are single-line.

Martin521 commented on src/Compiler/lex.fsl:
```diff
@@ -1080,6 +1082,13 @@ rule token (args: LexArgs) (skip: bool) = parse
      lexbuf.StartPos <- lexbuf.StartPos.ShiftColumnBy(n)
      HASH_IDENT(lexemeTrimLeft lexbuf (n+1)) }
 
+ | anywhite* ("#nowarn" | "#warnon") anystring
```
The directives were processed in the parser. This is no longer possible if we want to enable them everywhere (not only in top-level modules. Therefore I had to pull it forward into the lexer. (Commit 2 includes these changes).

Martin521 replied:
Multiline #nowarn directives don't exist in the whole of github except for the recently added test in this repo. So I guess it is safe to no longer support them.

Martin521 commented on src/Compiler/SyntaxTree/WarnScopes.fs:
```diff
@@ -0,0 +1,390 @@
+    /// a list of mapped sections (surrogate and original start lines).
+    type private LineMaps = Map<FileIndex, FileIndex * (LineNumber * LineNumber) list>
+
+    type private TempData =
```
I agree. I had actually made a note about it already :-)
I will change it.

Martin521 replied:
I renamed it.

auduchinok commented on src/Compiler/SyntaxTree/WarnScopes.fsi:
```diff
@@ -0,0 +1,32 @@
+open FSharp.Compiler.Text
+open FSharp.Compiler.UnicodeLexing
+
+module internal WarnScopes =
```
Any reason to make it internal?

We should somehow make the scopes information accessible to FCS users, given that the previous implementation is removed from SyntaxTree.

Martin521 replied:
The directives are available as trivia (now in `ParsedImplFileInputTrivia` / `ParsedSigFileInputTrivia`), see commit 5 of this PR.
But if more information is useful for tooling, please let me know what I should make available.

auduchinok replied:
Helpers like `IsWarnon` and `IsNowarn` are definitely useful.

Martin521 replied:
I made `IsWarnon` and `IsNowarn` public.
I would recommend, though, to rather rely on the filtering mechanism of the diagnostics loggers of the compiler (which use these two functions internally).

Martin521 commented on src/Compiler/Facilities/DiagnosticOptions.fs:
```diff
@@ -20,6 +20,8 @@ type FSharpDiagnosticOptions =
         WarnAsError: int list
         WarnAsWarn: int list
+        WarnScopesFeatureIsSupported: bool
+        mutable WarnScopeData: obj option
```
Yes, I did not really like to add a mutable field, but here are my "excuses".
1. I believe that the information about the warn scopes belongs into `FSharpDiagnosticsOptions` as much as all the other fields. _All these fields_ are used only in one single place in the whole compiler, namely in `PhasedDiagnostics.AddSeverity`. Passing the information separately does not fit its purpose and would add an additional parameter to many functions.
2. This information needs to be updated during postparse for every file. This is in the nature of the data: the #nowarn/#warnon directives change the way warnings are emitted.
3. I modeled this in some way after the `BufferLocalStore` of `lexbuf`. Just a way to store data for a certain aspect for the current compiler phase.
4. `WarnScopeData` is accessed only via private functions `WarnScopes.getWarnScopeData` and `WarnScopes.setWarnScopeData` in the `WarnScopes` module.

I will look into alternatives and see if I find something acceptable.
(For fsc, we could probably go back to tcConfigB and update the diagnosticsOptions there. But this may not work well for the other use cases. And it would not be safer or nicer or more idiosyncratic than the mutable field.)

Martin521 replied:
I have looked into alternatives, but I don't think there is an acceptable one. See also [this](https://github.com/dotnet/fsharp/pull/18049#discussion_r1934609273) comment for more details.

Note that, previously, `DiagnosticOptions` was formally immutable, but `tcConfig.diagnosticsOptions.WarnOff` was still mutated in an opaque and very complicated way after parsing (after each parse of a source file actually). I believe that making this explicit makes more sense.

psfinaki commented on src/Compiler/Service/FSharpCheckerResults.fs:
```diff
@@ -3219,10 +3219,6 @@ module internal ParseAndCheckFile =
-            let tcConfig =
-                ApplyNoWarnsToTcConfig(tcConfig, parsedMainInput, !! Path.GetDirectoryName(mainInputFileName))
-
             // update the error handler with the modified tcConfig
```
For my understanding, why this had to be removed (and nothing equivalent readded then)?

Martin521 replied:
The nowarn information (now: warn scope information) is no longer threaded from post-parse (where  it is created) to the type checking diagnostics logger by means of hiding it in the AST (where IMO it does not belong), but is stored now directly in the diagnostic options. See also the more detailed comments below.

(BTW I think the error handler update in the next line can also be removed. I will look into that.)

psfinaki replied:
Nice, thanks for the explanation!

psfinaki commented on vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.ErrorList.fs:
```diff
@@ -519,7 +519,7 @@ type staticInInterface =
 
     // In this bug, particular warns were still present after nowarn        
-    [<Fact>]
+    [<Fact(Skip="This needs to be re-enabled once VS is adapted to new WarnScopes")>]
```
I am afraid this is an issue for us. Compiler changes shouldn't degrade VS functionality. Can you share what's the problem appearing here? We can help with adjusting the things here.

Martin521 replied:
The above test creates a VS project and checks the reported errors. However, I am developing on Linux and have no access to VS. So, thanks for offering support here. Ideally, you would create a VS-independent test, working directly with the compiler service, for that same failing scenario. And I could take it from there.

This new feature necessitates changes in tooling. Not only for VS, but also for Rider, Ionide, Fantomas etc. I am happy to involve myself here. For the non-OSS tools this will necessarily be less hands-on than what I might be able to do for, say, Fantomas.

How is tooling affected by this PR?

First, for "coloring" (syntactically understanding) the #nowarn/#warnon directives themselves. Because of the new directive, a change in the respective API is unavoidable. As far as I understand the situation (see also the [diagram](https://github.com/Martin521/Review2412/blob/main/CompilerFlowChart.md) that I used in the recent [review](https://amplifyingfsharp.io/sessions/2024-12-13/)), Ionide and Fantomas get the necessary information out of the `Trivia` that are attached to the AST, while for VS additional artificial tokens are generated by `ServiceLexing` to carry the same information. This PR includes the necessary changes for both interfaces. But the consumption side in the tools, of course, must also be changed.

Secondly, the tools need the error/warning information that is influenced by the new directive. Ideally, nothing should change here. The filtering of the messages according to the various warnon/off defaults / compiler options / directives should be happening only internally in `CompilerDiagnostics.AdjustSeverity`. Only the way the necessary information flows into that method has changed. But the filtering should be the same for all use cases.

Why is the test failing (if the `Skip` is removed)?

The same source code is working fine in other test environments.
I suspect that VS is using the compiler service in a way that is not sufficiently covered by the existing Component and Service tests. But I have no way to find out, without VS.

psfinaki replied:
Alright, I will look into that and see what's happening there. If it's an easy fix, I'll push it here :)

psfinaki replied:
@Martin521 I think you've either protected the PR or your fork from changes by maintainers so I cannot push to your branch, hence created a PR with small test adjustments [here](https://github.com/Martin521/fsharp/pull/1).

Martin521 replied:
thanks

psfinaki reviewed (commented):
Hi @Martin521 and sorry this has been taking forever! I know you're patient but we shouldn't exercise your patience too much :)

We've discussed this PR within the team. In generally, we're in favor of having this in - it's a valuable improvement. That said, since the PR touches a few important places, we should be very sure that we don't compromise the compiler design and VS functionality. It's not that both are perfect but you know - the boy scout rule.

I added a few notes in accord with this - let's iterate here a bit and refine the changes.

Martin521 commented on src/Compiler/Facilities/DiagnosticOptions.fsi:
```diff
@@ -19,7 +19,9 @@ type FSharpDiagnosticOptions =
-      WarnAsWarn: int list }
+      WarnAsWarn: int list
+      WarnScopesFeatureIsSupported: bool
+      mutable WarnScopeData: obj option }
```
Yeah, I didn't particularly like it either.

Here are the reasons why I ended up with this mutable field.

1. The information about the warn scopes belongs into `FSharpDiagnosticsOptions` as much as all its other fields. _All these fields_ are used only in one single place in the whole compiler, namely in `PhasedDiagnostics.AddSeverity`. Passing the information separately does not fit its purpose and would add an additional parameter to many functions.
2. This information needs to be updated during postparse for every file. This lies in the nature of the data: the #nowarn/#warnon directives change the way warnings are emitted.
3. I modeled this in some way after the `BufferLocalStore` of `lexbuf`. Just a way to store data for a certain aspect for the current compiler phase.
4. `WarnScopeData` is accessed only via private functions `WarnScopes.getWarnScopeData` and `WarnScopes.setWarnScopeData` in the `WarnScopes` module.

I have looked into alternatives but didn't find anything acceptable.

One way would be to thread the `WarnScopeData` separately, or as part of the `inputs`, back to the "main" program. In the case of `fsc` that would be [here](https://github.com/dotnet/fsharp/blob/205a9fc3e36af193755139da30f9d503cceccaf1/src/Compiler/Driver/fsc.fs#L633). Bring it from there into `main2` and use it [here](https://github.com/dotnet/fsharp/blob/205a9fc3e36af193755139da30f9d503cceccaf1/src/Compiler/Driver/fsc.fs#L744), similarly to how it is done [today](https://github.com/dotnet/fsharp/blob/58560f8ce35a691415e10dd2b0189322d21afdfc/src/Compiler/Driver/fsc.fs#L749). This also means that the "merging" of the warn scopes of the different files has to happen here (rather than in the `WarnScope` module where it is much better placed). Next to `fsc`, this also has to happen (in even more complicated ways) for `fsi`, `BackgroundCompiler`, `TransparentCompiler` and `FSharpCheckerResults`.

Another way would be to do the above (threading of `WarnScopeData` through the function calls) only on the way up to the "main" program (i.e. `main1` in the case of `fsc`) and there do the merging, create a new `FSharpDiagnosticsOptions` with an immutable field `WarnScopeData`, and update `TcConfig` to contain this new version (by converting TcConfig to TcConfigB, adding it, going back to TcConfig, like in other places.)

But all of this is complicated and error prone, pulls the processing of the warn scope data into multiple places and is in the end doing the same unavoidable thing: changing the way that diagnostics is processed (which we would love to have immutable, but cannot) as a result of parsing the warn scope data, i.e. after parsing. (Which, BTW, you can still nicely see in the removed code of the previous comment (FSharpCheckerResults.fs:3223): in a nastily opaque way, the DiagnosticsOptions inside TcConfig were updated with the parse results. So much for immutability.)

To me, the way it is handled in the current PR version feels much cleaner and safer.

However, if the team is certain we should go another way, I will follow.

psfinaki replied:
Thanks for the detailed explanation. It's good that it follows at least some pattern then... Maybe @T-Gro you'll have ideas.

psfinaki replied:
Oh now I see Kevin asked the same in a sibling comment, sorry for this. I will resolve this comment then.

psfinaki commented:
Hi @Martin521. We've discussed this PR with the team yesterday. We are still on our low capacities but we don't want the PR to get stale, so we're thinking how to move on here.

I resolved a few comments which were duplicates or were addressed. I kept a few where we still have some "uncanny feelings". You know, not very happy and confident but also without an immediate idea on how to make things better. Also, some comments mention that you've wanted to look at other options, so please comment on those.

---

Let me ask you a few possibly annoying questions. Annoying because they might have popped up in some discussions already, but we want to summarize things a bit.

1) Do you consider any changes here breaking by any definition of breaking? (Fixing deep long-standing bugs might sometimes qualify...)

2) What parts of the compiler code are changes beyond propagation of the function parameters etc.? And how does the PR ensure that nothing breaks in that sense? Can you think of any testing there on top of what you've already added?

3) We respect the fact you don't have Visual Studio for testing the feature - that said, in principle, what do you think should be tested to be sure things work there and in other editors? I did some smoke testing yesterday in my VS (successfully) but I might have missed some scenarios. We might set up some sort of community session to try things out or brainstorm things a bit.

---

Thanks for your patience.

Martin521 commented:
> Let me ask you a few possibly annoying questions.

Nothing annoying here. I will tackle the questions and also the open comments one by one in the coming days.

Martin521 commented:
This PR has taken so long because I had to deal with messy parts of the compiler and a few times I had to change direction to work around an issue. But now we have a PR that cleans up a lot of complicated code, it implements the draft RFC flawlessly, and all existing and new tests succeed. Or so I thought until a week ago.

A week ago I found that some random test failure that appeared in some of the CI runs were no longer flaky tests that everybody experiences (like a few weeks ago), but are specific to this PR. The reason turned out to be that, for a certain class of tests, global resources (in this case, the `FileIndex` tables) are kept across tests but are used in ways that were not foreseen (in this case, using different source files with the same name). Therefore these tests are not suited for the parallelization that was recently introduced. It seems that for other uses of `FileIndex` (like with #line directives) this issue is not visible or just very rare, because of low test coverage.

A fix would be to make sure that in testing (for tests that use the same `checker` instance) the file names are unique. But there are some 300 tests that explicitely use "test.fs" or "test.fsx". (Pinging @majocha here for possibly other ideas.)

In none of the production scenarios this problem is relevant, but if we want to keep test parallelization (which I guess we want), we have to either fix the test naming problem, or I have to again look for another way to enable "scoped nowarn".

I will therefore postpone answering the [questions](https://github.com/dotnet/fsharp/pull/18049#issuecomment-2633699385) until I have found a new solution (which might make the questions obsolete) or until I finally have to give up.

Martin521 commented:
In any case, however, we need to get agreement on the main compatibility question. That's the question if we should remove the current inconsistency between Fsc and the IDEs regarding `#nowarn`, or deepen it.
I.e. if we should introduce a breaking change to make them consistent, or enable "warn scopes" only for fsc.

This is a question that should be [discussed](https://github.com/fsharp/fslang-design/discussions/786#discussioncomment-10716719) with the [RFC](https://github.com/fsharp/fslang-design/pull/782), but here is the technical / historical background that gives rise to the question.

The `#nowarn` directive is handled in two very different ways. One way (my compiler archeology lets me think this is the older one) is to search through the modules of each parsed file, to extract the directives (but only for top-level modules) and for each #nowarn directive to either add a `--nowarn` flag to the compiler options (this is done in the checker when handling projects, and in fsi in certain situations) or to fake a `--nowarn` flag by mutating (in an obfuscated way) the "immutable" `tcConfig.diagnosticsOptions.WarnOff` (this is done in other situations in the checker and fsi, and when compiling scripts).

It seems that at some point then somebody wanted to create a better design. They created the field `ScopedPragmas` of the parsed input. Here, the `#nowarn` directives are collected *with their file index and line number*, so that the effect of the #nowarn can be scoped to the lines following the directive (until eof). These `ScopedPragmas` are then used in creating a `DiagnosticsLoggerFilteringByScopedPragmas`, which in principle can filter the warnings in the right "scoped" way. But then the scoping is disabled again in that logger except when it is called from `fsc`. Also, all the flag based #nowarn processing was left in place except for the fsc case. Because of the extremely complicated and obfuscated way the diagnostics loggers are created and used, it is almost impossible to find out which of the two ways of `#nowarn` processing is used where. Sometimes they are even just used in parallel (like [here](https://github.com/dotnet/fsharp/blob/ee559975b208d5eb64cc9c3c832a29de1bf5f131/src/Compiler/Service/TransparentCompiler.fs#L1394C1-L1402C1)). So, that nice new design was implemented only for the "fsc" path and the rest was left in an inconsistent state. (BTW all of this happened before this repo was created, so nobody to "blame".)
  
So, there are now inconsistent ways in which `#nowarn` works. In straight fsc compilation of .fs and .fsi files, it works "scoped", i.e. is in effect from the directive to eof. But in your IDE, that same `#nowarn` is in effect *for the whole file*.  And if you include a script file in your project and compile the project, a `#nowarn` in the script (or any script #loaded by the first one) is in effect *for the whole compilation*.

So, assuming we find a way to implement "warn scopes", should we implement it only for fsc, or everywhere?

majocha commented:
> A fix would be to make sure that in testing (for tests that use the same `checker` instance) the file names are unique. But there are some 300 tests that explicitely use "test.fs" or "test.fsx". (Pinging @majocha here for possibly other ideas.)

Probably hardcoded default names without a directory? Like this:
https://github.com/dotnet/fsharp/blob/b64ea2b04b729ef02ca79b4ad85bedc45d26679b/tests/FSharp.Test.Utilities/CompilerAssert.fs#L1023
I'll try to replace the occurrences with generated names, shouldn't take long.

Martin521 commented:
> I'll try to replace the occurrences with generated names, shouldn't take long.

If that's doable - that would be really great.

psfinaki commented:
@majocha - thanks for your relentless support here :) Indeed, this would be great.

---

@Martin521 thanks for the detailed answer. I think I understand the inconsistency. We might want to discuss it a bit, but before that - what do you think is easier to do from the implementation perspective? 

Very generally speaking, the preference is:
consistent in a wrong way << inconsistent << consistent in the right way :)

Martin521 commented:
> what do you think is easier to do from the implementation perspective?

From an implementation point of view, a consistent implementation of warn scopes for all use cases is definitely simpler. It is simpler to implement and also enables us to remove all the complicated code around the old way of dealing with #nowarn.

(If you want to have a taste of the code around the #nowarn collection that I mentioned above, have a look at `ApplyMetaCommandsFromInputToTcConfigAndGatherNoWarn`, which calls `ProcessMetaCommandsFromInput`. The latter looks as if somebody has just found out that functions can be passed around. "Functional" design at its worst. No wonder that [here](https://github.com/dotnet/fsharp/blob/7866d665f47e1dc650248b3d8813d5c99a6c5d1a/src/Compiler/Driver/fsc.fs#L659) the author no longer knows what they are doing (the comment says "apply any nowarn flags", but the code doesn't do that at all.)

Martin521 commented:
> @Martin521, so a consistent and simpler approach would be preferred here. In general, we count with breaking changes here.
> 
> Another way to put it - when users start relying on bugs, fixing bugs becomes a breaking change. But often that's still the right way to go. We're currently having such discussion [here](https://github.com/dotnet/fsharp/issues/18298). But it might require some communication and documentation later :)

Thanks, that's very clear and I am glad you see it that way.

First, however, I have now to dive still deeper into the uses of the global `FileIndex`. I am relying on it now for scoped nowarn, and that is fine for all production scenarios and all sequential testing, but not for parallel testing. It is not so easy to decide if that is a test suite issue, or if I should not have used `FileIndex` in the way that I do. I want to really understand and document that in detail (which will probably take a while) and then come back to your 3 questions.

psfinaki commented:
Sure, let's tackle issues one by one :)

Martin521 commented:
I have fixed the testing issue mentioned above.

For those interested, here are some details.

FileIndex.fileIndexTable is a global mutable resource that represents a bidirectional mapping between file paths and file indices.
Originally, FileIndex was probably only meant to store source file identity in a compact way in `range` structs, and the FileIndexTable was used to get back to the file path. But then additional uses crept in. For example:
- `DiagnosticsLoggerFilteringByScopedPragmas` [uses the file index](https://github.com/dotnet/fsharp/blob/c281bd08b2d6bcbf60f5ccaa925f17de8ef714f1/src/Compiler/Driver/CompilerDiagnostics.fs#L2315) to match warning ranges against #nowarn directives that belong to the same file index.
- Graph type checking uses FileIndex extensively to [identify](https://github.com/dotnet/fsharp/blob/c281bd08b2d6bcbf60f5ccaa925f17de8ef714f1/src/Compiler/Driver/ParseAndCheckInputs.fsi#L46) a "physical file in the current project".
- And now I want `WarnScopes` to use FileIndex to match warning ranges against warn scopes ranges that belong to the same file index.

This is all fine in any production scenario (fsc, fsi, FSAC), because all files have unique paths. Mapping to FileIndex and back is not a problem. In Testing, however, there are tests that use, in different tests, the same file path "test.fs" in what looks to the compiler service like the same project ("test.fsproj"). When these tests run in parallel, matching ranges by file index is no longer reliable because the same file index might now be used by different tests and a, say, #nowarn directive from one test might be matched against a warning range of another test.

What were the options?
- Either keep testing as is, but avoid in the compiler any use of FileIndex other than getting the file path (avoid, for example, the use of `range.FileIndex` to find out if two ranges belong to the same file),
- Or require that in testing all file paths are distinct,
- Or require that in testing all file paths in a project are distinct (where "project" is any collection of tests that end up using the same tcConfig).

I went for the third option. Because it was simple to implement on the testing side (thanks to @majocha for providing the code), while implementing warn scopes under the first option would be very difficult.

Martin521 commented:
Now, finally, the answers to the [questions](https://github.com/dotnet/fsharp/pull/18049#issuecomment-2633699385).

> 1. Do you consider any changes here breaking by any definition of breaking? (Fixing deep long-standing bugs might sometimes qualify...)
 
This is completely covered by the [compatibility section](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1146-scoped-nowarn.md#compatibility) of the RFC.

(Note that, currently, the code in this PR produces the new behavior for scripts (item e) independent of the feature flag. If we want to keep the old behavior for old language versions, this still needs to be fixed.)



> 2. What parts of the compiler code are changes beyond propagation of the function parameters etc.? And how does the PR ensure that nothing breaks in that sense? Can you think of any testing there on top of what you've already added?

The substantial changes in this PR are all based on one of two reasons.
One is the fact that the warn directives can no longer be collected during parsing (since they can now appear not only in top-level modules, but anywhere). So - in the absence of a preprocessor - we have to collect them now during lexing, similar to the processing of #if/#else/#endif directives. The related code changes are in the second commit of this PR.

The second reason is the feature itself and the processing of the warn directives that it needs. Previously, the #nowarn directives were processed in a number of different places (postparse, fsc, fsi, bgCompiler, TransparentCompiler, FSharpCheckerResult, ScriptClosure) to make them land in either `FSharpDiagnosticOptions.WarnOff` or in a `--warnoff` compiler option, often in complicated ways that were not easy to follow and IMO already crying for refactoring. Now, with the additional complexity of the Warn Scopes feature (we have to compute "on" and "off" scopes from the #nowarn/#warnof directives for each warning number and file), that way of handling the directives was no longer maintainable. So, all that complicated code was removed and instead we have now all warn directive processing in a single module `WarnScopes` with a very small surface. The related changes to existing code are in commits 3 and 4 of this PR.

Every large refactoring comes, of course, also with risks. I want to put these risks here into two categories.
The first one is about the new implementation deviating from the spec (the RFC). I believe the risks in this category are low. The whole feature functionality is implemented in the single `WarnScopes` module, is accessed only internally from very few places, and is quite well tested in `FSharp.Compiler.ComponentTests:CompilerDirectives.NoWarn`.

The second category is about incompatibilities with the previous behavior.
Next to the known and intended incompatibilities listed in the RFC, there might still exist unknown incompatibilities. Due to the previously distributed (duplicated) processing of #nowarn in the many different "compilers" (fsc, fsi script, fsi eval, bg, transparent, checker), this is very difficult to track. I also suspect that in general not all of these paths are sufficiently covered by tests. But I don't think the risk is too high - nothing popped up when building this big repo.



> 3. We respect the fact you don't have Visual Studio for testing the feature - that said, in principle, what do you think should be tested to be sure things work there and in other editors? I did some smoke testing yesterday in my VS (successfully) but I might have missed some scenarios. We might set up some sort of community session to try things out or brainstorm things a bit.

This is a very good question.

Currently, I cannot test any of the IDEs.
As far as I understand, there is a way to make a custom version of FsAutoComplete with my compiler version, and to configure Ionide to use it. But will have to find out if and how this works in my dev container setup.

If I knew how exactly the IDEs use the compiler service to obtain diagnostics, I would add tests for exactly these scenarios to FSharp.Compiler.Service.Tests.

It might also make sense to add a few warn scope tests to vsintegration and/or FsAutoComplete.

To have a session with some experts is a good idea.

psfinaki commented:
Summoning @baronfel for FSAC thoughts :)

psfinaki commented:
So I think we are getting there. I'll give another fresh review to the compiler changes and we need to come up with editor testing scenarios.

We can have a community for this to just brainstorm things here once again.

psfinaki commented:
Sorry I pressed the wrong button!

Martin521 commented:
Looking into FsAutoComplete, it seems that `NotificationEvent.ParseError` is triggered only from `AdaptiveState.parseAndCheckFile`, which in turn gets the diagnostics from `ParseAndCheckFileInProject` of the compiler service.
So, assuming that is the only relevant event, I should add warn scope tests, using `FSharpChecker.ParseAndCheckFileInProject`, to `FSharp.Compiler.Service.Tests`.

psfinaki commented:
@Martin521 yeah this is definitely a good starting point there :)

Martin521 commented:
Ok, I have added various compiler service and AST tests that should safeguard the editor behavior.
The three unsuccessful CI checks are obviously failing for reasons not related to this PR.
I would now declare this PR ready for review.

T-Gro commented:
> Ok, I have added various compiler service and AST tests that should safeguard the editor behavior. The three unsuccessful CI checks are obviously failing for reasons not related to this PR. I would now declare this PR ready for review.

This is great news @Martin521!  I will set aside some time during April to review this fully.

T-Gro commented:
Hi @Martin521 , I am reviewing + testing this branch incl. editor scenarios today.
One thing I wanted to do as a test exercise is to visit existing code locations with a file-wide "nowarn", delete them, and apply nowarn+warnon locally.

However, when I removed the top level old-style "nowarn", nothing happened - no warnigs!

I then had to make some other tweaks, such as adding a pair of "nowarn"+"warnon" somewhere in the file, for the actions to really happen. Once I broke this initial staleness, it all worked fine from there.


So there is this special scenario where the removal of a single old-style "nowarn" does not do anything.
Since this might be the first introduction scenario for many (replacing old-style "nowarn" with scoping), would you mind taking a look if there could be a caching issue or other form of error causing this?

Video attached:

https://github.com/user-attachments/assets/e40b24cb-3878-4616-9506-2bcaed12c334

T-Gro commented:
More focused video on the same problem, it might give you some hints on what triggers a change perhaps?

https://github.com/user-attachments/assets/d2236929-3e07-4af6-abd0-7ac2a4f9ddcb

Martin521 commented:
Thanks for testing! I actually came across a similar "startup" issue when testing with Ionide a few days ago. I have an idea what the cause might be and will find out. It will take a few days though since I am travelling.

Martin521 commented:
It seems that due to my initial focus on the compile (fsc) use case I didn't cover all editor use cases properly. It won't be so difficult to fix this, but it will take some time. Therefore, I will make this PR draft again. Sorry for prematurely asking for review.

T-Gro commented:
No need to apologize :).
I would suspect the changes around removal of  `ApplyNoWarnsToTcConfig` in  FSharpCheckerResults. I know that things are done differently after this PR, but there is probably something causing the diagnostic chain to get stale.

Martin521 commented:
I added tests now also for multiple edit / typecheck cycles of the same file; fixed a flaw in the warn scope storage; fixed TransparentCompiler where in a certain situation it didn't play well with warn scopes; got FSAC/Ionide working with warn scopes and started explorative testing. So, I am submitting this once more for review.

![Nowarn](https://github.com/user-attachments/assets/7845555a-4b88-45d3-aca4-921ebfb59715)

T-Gro commented:
Thanks @Martin521 .
Was any VS testing done after the change?

If not, I will schedule myself for later to test this change applied to VS and check the behavior of the incremental-compilation.
(I already apologize, it can take a few weeks until I get to that - if you can find any other volunteer to test this in VS, I am fine with another evidence (e.g screen recordings))

Martin521 commented:
> Was any VS testing done after the change?
> 
> If not, I will schedule myself for later to test this change applied to VS and check the behavior of the incremental-compilation. (I already apologize, it can take a few weeks until I get to that - if you can find any other volunteer to test this in VS, I am fine with another evidence (e.g screen recordings))

No, I don't have VS installed.
Any volunteers out there?

Martin521 commented:
> Since the RFC does have a breaking chance section, I decided to pick this PR as our first tester of the new SDK breaking chance policy. You do not have to do anything now, I will test how the labels help us track which doc changes are needed and if the process for documenting changes labelled as potentially breaking works smoothly.

Let me know I should create a new issue, using the breaking change template, with a pointer to the old one, and link this PR to it.

majocha commented:
> No, I don't have VS installed. Any volunteers out there?
I just built from this branch and installed the extension in VS.
Looks like basics work fine:
![image](https://github.com/user-attachments/assets/c6aa56d1-d933-41a1-bf14-0df69fa891f0)

I'll keep using this with an eye on any potential problems.

T-Gro commented:
> I'll keep using this with an eye on any potential problems.

@majocha : Did you test some scenarios where incremental compilation is needed? (e.g. removing an existing `nowarn`, did warnings immediately pop out? )

If yes and it worked (it failed here, see my screencast above), we can accept that as last test evidence and get this in.

majocha commented:
> > I'll keep using this with an eye on any potential problems.
> 
> @majocha : Did you test some scenarios where incremental compilation is needed? (e.g. removing an existing `nowarn`, did warnings immediately pop out? )
> 
> If yes and it worked (it failed here, see my screencast above), we can accept that as last test evidence and get this in.

Yes it works now. I tested with some edits, also wrapping a #nowarn in #if DEBUG etc. Warnings do show up and go away.

T-Gro commented:
Thanks @majocha  for spotting this, it is somewhat similar to what I have experienced - some stale config data in the incrementalbuilder.
I also see that this PR explicitly deleted code which tried to maintain this information, so I am afraid it still might not be fully correct when it comes to design-time experience.

T-Gro commented on src/Compiler/Service/TransparentCompiler.fs:
```diff
@@ -1392,14 +1391,10 @@ type internal TransparentCompiler
                         tcConfig.flatErrors
                     )
 
-                // Apply nowarns to tcConfig (may generate errors, so ensure diagnosticsLogger is installed)
```
Or here , alternatively

T-Gro commented:
> I too feel uneasy about diagnosticOptions beeing mutable now. For example, it seems they're indirectly included as part of cache key in TransparentCompiler. There is potential for subtle and hard to track issues.
> 
> I also wonder if ConditionalWeakTable could be another way to avoid directly threading that data across function calls.

`ConditionalWeakTable` keyed with snapshot of the source file (ISourceText ?), or something else?
The issues so far manifest under code-edit scenarios, using stale / out-of-date diagnostic options worries me the most now.

majocha commented:
> I thought ParsedInput. That's where the data is created, so should stay in sync.

This won't work, because in `AdjustSeverity` where the data is consumed, the only thing we have is the `range` of `PhasedDiagnostic`.

Martin521 commented:
First, thanks to @T-Gro and @majocha for testing and reviewing this.
Too bad that there are still mistakes in code-edit scenarios. The reason could be in my code, but I bet it is in some inconsistency in BackgroundCompiler/TransparentCompiler/IncrementalBuild. I have come across a few of them already (e.g. #18366, #18380, and one that I fixed in a recent commit to this PR).

However, I do understand your reluctance to accept code that is sensitive to certain constellations that were not problem so far but are seen as "inconsistency" by my code. (And also, I don't know how difficult it will be to find a repro for the mistake in order to fix it.)

I therefore want to try once more to find an alternative to the tcConfig.diagnosticsOptions solution. I have an idea that I need to check.

To still answer the question of @T-Gro about `ApplyNoWarnsToTcConfig`. This function was used for scripts and by `service`, but not in `fsc` (except for scripts). It just added any #nowarn to tcConfig.diagnosticsOptions.Warnoff, thus making it valid for the whole file. In `fsc`, a different mechanism was used that made #nowarn valid only after the directive. We have agreed in the RFC to change this and have the same "scoped" behavior for fsc, service and scripts. Therefore, all `ApplyNoWarnsToTcConfig` (and equivalent functions) were removed and the `fsc` mechanism (adjusted to WarnScopes) is now used for all cases.

Anyway, my ask is for some time to check the alternative that I mentioned above.

majocha commented:
@Martin521, 
I can repro this only with TransparentCompiler. The fact that it caches diagnostics may have something to do with it.
This is probably not very helpful, but, anyway, what I did was toggle comment / uncomment on `#warnon 25` in the following sample:
```
module A
#nowarn 25
match None with None -> ()
#warnon 25
let x =
    let a = match None with Some () -> 1 
    match None with None -> 1 
match None with Some () -> () 
match None with None -> ()  
```
After few tries I was able to get it stuck in a wrong state. Making some edit that is not yet cached makes it work again.

majocha commented:
Again this is with TransparentCompiler:
https://github.com/user-attachments/assets/f0cf8e06-58f7-4a0a-8164-883338e948bb

Martin521 commented:
Thanks!

majocha commented:
I just noticed `#if DEBUG` in code sometimes goes out of sync, too. I wonder if it's the same bug?

Martin521 commented:
It seems in TransparentCompiler the file version part of the cache key is not reliably updated on edits. But I have not yet found the place where it goes wrong.

Martin521 commented:
The above issue happens only in a certain situation in interactive (IDE) use and only when `TransparentCompiler` (TC) is enabled. If an edit of a warn directive makes a warning appear (or disappear) and you then revert that edit, the warning does not disappear (or appear, resp.) as you would expect.

The last commits that I have pushed should fix this (hopefully last) issue.

Here is the long story for those interested.

I first have to state again why the warn scope data are stored in `tcConfig.diagnosticOptions`.

Following the RFC, we want the warn directives to behave consistently across all compiler modes, i.e. in fsc-compiled .fs/.fsi files, in fsc-compiled .fsx files, in the IDEs (compiler service), and in fsi compilation (both scripts and interactive).

This means we have to replace two existing mechanisms (one in fsc and one for the other cases) by a single new one. (The fsc mechanism stores the #nowarn information during post-parse in `ParsedInput` and then brings it during type checking into a special diagnostics logger, which makes it available to `AdjustSeverity`. In all other cases, the information is collected from `ParsedInput` and is put into `tcConfig.diagnosticOptions`, which is available in `AdjustSeverity`.)

To use the current fsc mechanism also for the other cases turns out to be impossible. At least not before a BIG cleanup of the whole diagnostics logging and background compiler / fsi. The call hierarchies leading to post-parse and to `AdjustSeverity` are so deep and so broad, that threading something from one to the other is an impossible task (except for the fsc case). So, the only way left is to follow the other path and store the warn scope data for the project in `tcConfig.diagnosticOptions`.

This has been implemented and works nicely everywhere *except* for the situation mentioned above. Here is what's happening. The  TC caches parse and check results. Caching is based on keys. For the parse cache, the key is (if we simplify the matter a little) just the file name and a hash of the file content (source text). If during editing the content changes, the file is parsed again (including update of the warn scope data in the project's `tcConfig.diagnosticOptions`) and the new parse result is cached. If for some reason the editor asks for a check of the file while the content has not changed, the results are taken from the cache and the warn scope data stays the same, which is fine. If, however, during editing, we return to a content state that has been parsed before, we get from the cache the previous parse result, but the warn scope data is not updated. This explains the issue that we have observed in exploratory testing of warn scopes in the IDEs.

What can we do now?

a) Postpone "scoped nowarn" (beyond F# 10) and first do a lot of cleanup. Remove a lot of the code duplication in the many compilation paths. Unify/simplify diagnostics logging for the different paths. Finish TC (bring it to production quality) and remove the old background compiler. Replace the current hack that implements #line directives by something sensible. Etc etc. (TBH, in hindsight, we should never have started "scoped nowarn" without the cleanup.)

b) Fix the issue by disabling parse result caching. This is a serious and simple option. I don't think it will affect performance in a significant way. Parsing takes little resources compared to type checking. (Also, ironically, `ParseAndCheckFileInProject` (the main entry point for the editor that leads to parsing) first calls the caching parser, then does the type checking, then does an unconditional parse (not chached) of the same file again (via `ComputeScriptClosure`).)

c) Fix the issue by making sure the parse cache key is changed (incremented) with every edit. This would be a fix with zero performance impact. But implementing it requires a quite deep redesign of the TC.

I have implemented b) now, leaving a) and c) to a next step.
Please let me know if there are other insights, opinions or advice.

T-Gro commented:
Thanks for the explanation @Martin521.

If I understood it correctly, caching non-changing data is fine.
If a single change happens, we are also fine if re-parsing happens.

Which means that the only way problem can appear is if there are 2+  entries in the parsing cache, for the same file. (one current and one "to go back").

I do not like the approach to fully disable the parsing cache - there are tooling needs also for parsed results and we do have a tooling layer which simply assumes content is cached if it is not changing. (i.e. tooling happily asks for parsed results multiple times for various operations).


Could you please elaborate any possible downsides of an approach b2:
- Keep the parsing cache, but modify it so that it cannot go back in time. i.e. it would only keep 1 current parsed result for file. If stuff changes, it is reparsed. If tooling asks dozens of times for the same file, cache result is returned.


Would there be any downsides then?

Martin521 commented:
> Let me look into "b2" (size 1 cache), it sounds like a very good idea.

Making the cache single-version was easier than I thought ([here](https://github.com/dotnet/fsharp/blob/5a436aa68f707b8b96c441f8168373f482e95590/src/Compiler/Service/TransparentCompiler.fs#L304)).
The difficult part was to create a [regression test](https://github.com/dotnet/fsharp/blob/32db9537fdaa61fc72d4873469876095901ccc88/tests/FSharp.Compiler.Service.Tests/WarnScopeTests.fs#L180) that would duplicate the interactive behavior that we observed.

The CI MacOS issue doesn't seem to be related to this PR. In any case, I have no means to find out.

T-Gro reviewed (approved):
Thank you @Martin521 for your patience with fixing the outstanding IDE scenarios!
This is a big contribution and will improve the capabilities for handling warnings in F# codebases!

Martin521 commented on src/Compiler/SyntaxTree/SyntaxTrivia.fsi:
```diff
@@ -44,9 +44,9 @@ type CommentTrivia =
+/// Represents additional information for ParsedInput
 [<NoEquality; NoComparison>]
-type ParsedImplFileInputTrivia =
+type ParsedInputTrivia =
```
While looking into the necessary changes in Fantomas to accomodate this PR, I found that unifying these two types simplifies a lot of code in both the compiler and fantomas.core.

Martin521 commented:
@T-Gro 
Can this be merged? Or is there anything I can still do here?

T-Gro commented:
> @T-Gro Can this be merged? Or is there anything I can still do here?

It can be merged, now just waiting for a second tick from other maintainers. I asked offline to give it, expect it to be merged soon.

Thorium commented:
987 files changed, a huge thing! :-D
Now that the WarnDirectives are carried across, would it be possible for TypeProviders to raise warnings in addition to errors?

kerams commented:
@Thorium, https://github.com/dotnet/fsharp/pull/14483 - there was not interest in it

psfinaki commented:
Well done @Martin521! Your persistence here was exemplary - great job :)
